File: mcp-server/formatters.py
#!/usr/bin/env python3
"""
MemOS MCP Server Formatters Module

Contains data formatting functions for displaying memory search results and graphs.
"""

import re

from typing import Any

from pydantic import BaseModel


class MemorySearchResult(BaseModel):
    """Structured memory search result."""
    id: str
    content: str
    relevance: float = 1.0
    metadata: dict[str, Any] = {}


def format_memories_for_display(data: dict) -> str:
    """Format memory search results for readable display."""
    results = []

    # Process text memories
    text_mems = data.get("text_mem", [])
    for cube_data in text_mems:
        cube_id = cube_data.get("cube_id", "unknown")
        memories_data = cube_data.get("memories", [])

        # If memories is a dict with nodes (tree_text mode), extract nodes
        memories = []
        if isinstance(memories_data, dict) and "nodes" in memories_data:
            memories = memories_data["nodes"]
        elif isinstance(memories_data, list):
            memories = memories_data

        if memories:
            results.append(f"## 📦 Cube: {cube_id}")
            results.append("")

            # Group by type
            grouped = {}
            for mem in memories:
                memory_text = mem.get("memory", "")
                # Try to extract type from [TYPE] prefix
                type_match = re.match(r"^\[([A-Z_]+)\]", memory_text)
                mem_type = type_match.group(1) if type_match else "PROGRESS"
                if mem_type not in grouped:
                    grouped[mem_type] = []
                grouped[mem_type].append(mem)

            # Display by type
            for mem_type, items in grouped.items():
                results.append(f"### 🏷️ Type: {mem_type}")
                results.append("")

                for i, mem in enumerate(items, 1):
                    memory_text = mem.get("memory", "")
                    mem_id = mem.get("id", "")  # Full UUID for delete operations

                    # Remove the [TYPE] prefix from display text if present
                    display_text = re.sub(r"^\[[A-Z_]+\]\s*", "", memory_text)

                    # Extract first line as title
                    first_line = display_text.split("\n")[0][:100]
                    if len(display_text.split("\n")) > 1 or len(display_text) > 100:
                        results.append(f"#### {i}. {first_line}")
                    else:
                        results.append(f"#### {i}. {display_text}")

                    results.append(f"ID: `{mem_id}`")
                    results.append("")

                    # Detect if it's a code block (simple heuristic)
                    if "```" not in display_text and any(
                        line.strip().startswith(("import ", "def ", "class ", "export ", "const ", "let ", "var "))
                        for line in display_text.split("\n")
                    ):
                        results.append("```python")
                        results.append(display_text)
                        results.append("```")
                    else:
                        results.append(display_text)

                    results.append("")
                    results.append("---")
                    results.append("")

    if not results:
        return "No memories found matching your query."

    return "\n".join(results)


def format_graph_for_display(data: list) -> str:
    """Format knowledge graph results with relationships for readable display."""
    results = []

    for cube_data in data:
        cube_id = cube_data.get("cube_id", "unknown")
        memories_list = cube_data.get("memories", [])

        if not memories_list:
            continue

        results.append(f"## 🧠 Knowledge Graph: {cube_id}")
        results.append("")

        for mem_data in memories_list:
            # Extract nodes and edges
            nodes = mem_data.get("nodes", [])
            edges = mem_data.get("edges", [])

            # Build node lookup for relationship display
            node_lookup = {}
            for node in nodes:
                node_id = node.get("id", "")
                node_memory = node.get("memory", "")
                # Clean up memory text for mermaid (remove newlines and special chars)
                clean_text = node_memory.replace("\n", " ").replace('"', "'").replace("[", "(").replace("]", ")")
                node_lookup[node_id] = clean_text[:50] + "..." if len(clean_text) > 50 else clean_text

            # Display nodes
            if nodes:
                results.append("### [NOTE] Memory Nodes")
                results.append("")
                for i, node in enumerate(nodes[:10], 1):  # Limit to 10 nodes
                    memory = node.get("memory", "")
                    first_line = memory.split("\n")[0][:100]
                    node_id = node.get("id", "")  # Full UUID for delete operations
                    results.append(f"{i}. **{first_line}**")
                    results.append(f"   ID: `{node_id}`")
                    results.append("")

            # Display relationships with Mermaid diagram
            if edges:
                results.append("### 📊 Relationship Diagram (Mermaid)")
                results.append("")
                results.append("```mermaid")
                results.append("graph TD")

                # Style definitions
                results.append("    classDef cause fill:#f96,stroke:#333,stroke-width:2px;")
                results.append("    classDef relate fill:#bbf,stroke:#333,stroke-width:1px;")
                results.append("    classDef conflict fill:#f66,stroke:#333,stroke-width:2px,stroke-dasharray: 5 5;")

                added_edges = set()
                for edge in edges:
                    source_id = edge.get("source", "")
                    target_id = edge.get("target", "")
                    rel_type = edge.get("type", "UNKNOWN")

                    # Skip PARENT relationships, only show semantic ones
                    if rel_type == "PARENT":
                        continue

                    # Avoid duplicate edges in diagram
                    edge_key = f"{source_id}-{target_id}-{rel_type}"
                    if edge_key in added_edges:
                        continue
                    added_edges.add(edge_key)

                    source_text = node_lookup.get(source_id, source_id[:8])
                    target_text = node_lookup.get(target_id, target_id[:8])

                    # Sanitize IDs for mermaid (must be alphanumeric)
                    s_id = f"node_{source_id[:8]}"
                    t_id = f"node_{target_id[:8]}"

                    # Format relationship in mermaid
                    if rel_type == "CAUSE":
                        results.append(f'    {s_id}["{source_text}"] -- CAUSE --> {t_id}["{target_text}"]:::cause')
                    elif rel_type == "RELATE":
                        results.append(f'    {s_id}["{source_text}"] -. RELATE .- {t_id}["{target_text}"]:::relate')
                    elif rel_type == "CONFLICT":
                        results.append(f'    {s_id}["{source_text}"] == CONFLICT == {t_id}["{target_text}"]:::conflict')
                    elif rel_type == "CONDITION":
                        results.append(f'    {s_id}["{source_text}"] -- CONDITION --> {t_id}["{target_text}"]')
                    else:
                        results.append(f'    {s_id}["{source_text}"] -- {rel_type} --> {t_id}["{target_text}"]')

                results.append("```")
                results.append("")

                # Textual fallback for terminals that don't render mermaid
                results.append("### [LINK] Textual Relationships")
                results.append("")
                results.append("```")
                for edge in edges:
                    if edge.get("type") == "PARENT":
                        continue
                    s_text = node_lookup.get(edge.get("source"), "???")[:40]
                    t_text = node_lookup.get(edge.get("target"), "???")[:40]
                    results.append(f"[{s_text}] --{edge.get('type')}--> [{t_text}]")
                results.append("```")
                results.append("")

        results.append("---")

    if not results:
        return "No memories or relationships found."

    return "\n".join(results)

File: mcp-server-node/src/formatters.ts
/**
 * MemOS MCP Server Formatters
 *
 * Formats memory search results and knowledge graphs for display.
 */

import type { SearchData, MemoryNode, GraphEdge } from "./types.js";

// ============================================================================
// Memory Display Formatter
// ============================================================================

export function formatMemoriesForDisplay(data: SearchData): string {
  const results: string[] = [];

  const textMems = data.text_mem ?? [];
  for (const cubeData of textMems) {
    const cubeId = cubeData.cube_id ?? "unknown";
    const memoriesData = cubeData.memories;

    let memories: MemoryNode[] = [];
    if (memoriesData && !Array.isArray(memoriesData) && memoriesData.nodes) {
      memories = memoriesData.nodes;
    } else if (Array.isArray(memoriesData)) {
      memories = memoriesData as MemoryNode[];
    }

    if (memories.length === 0) continue;

    results.push(`## 📦 Cube: ${cubeId}`);
    results.push("");

    // Group by type
    const grouped: Record<string, MemoryNode[]> = {};
    for (const mem of memories) {
      const memText = mem.memory ?? "";
      const typeMatch = memText.match(/^\[([A-Z_]+)\]/);
      const memType = typeMatch ? typeMatch[1] : "PROGRESS";
      if (!grouped[memType]) grouped[memType] = [];
      grouped[memType].push(mem);
    }

    // Display by type
    for (const [memType, items] of Object.entries(grouped)) {
      results.push(`### 🏷️ Type: ${memType}`);
      results.push("");

      for (let i = 0; i < items.length; i++) {
        const mem = items[i];
        const memText = mem.memory ?? "";
        const memId = mem.id ?? "";

        // Remove [TYPE] prefix
        const displayText = memText.replace(/^\[[A-Z_]+\]\s*/, "");
        const firstLine = displayText.split("\n")[0].slice(0, 100);

        if (displayText.split("\n").length > 1 || displayText.length > 100) {
          results.push(`#### ${i + 1}. ${firstLine}`);
        } else {
          results.push(`#### ${i + 1}. ${displayText}`);
        }

        results.push(`ID: \`${memId}\``);
        results.push("");

        // Detect code blocks
        const hasCodeIndicator = displayText.split("\n").some((line) =>
          /^(import |def |class |export |const |let |var )/.test(line.trim())
        );
        if (!displayText.includes("```") && hasCodeIndicator) {
          results.push("```python");
          results.push(displayText);
          results.push("```");
        } else {
          results.push(displayText);
        }

        results.push("");
        results.push("---");
        results.push("");
      }
    }
  }

  if (results.length === 0) {
    return "No memories found matching your query.";
  }

  return results.join("\n");
}

// ============================================================================
// Graph Display Formatter
// ============================================================================

export interface GraphCubeData {
  cube_id?: string;
  memories?: Array<{
    nodes?: MemoryNode[];
    edges?: GraphEdge[];
  }>;
}

export function formatGraphForDisplay(data: GraphCubeData[]): string {
  const results: string[] = [];

  for (const cubeData of data) {
    const cubeId = cubeData.cube_id ?? "unknown";
    const memoriesList = cubeData.memories ?? [];

    if (memoriesList.length === 0) continue;

    results.push(`## 🧠 Knowledge Graph: ${cubeId}`);
    results.push("");

    for (const memData of memoriesList) {
      const nodes = memData.nodes ?? [];
      const edges = memData.edges ?? [];

      // Build node lookup
      const nodeLookup: Record<string, string> = {};
      for (const node of nodes) {
        const nodeId = node.id ?? "";
        const nodeMemory = node.memory ?? "";
        const cleanText = nodeMemory.replace(/\n/g, " ").replace(/"/g, "'").replace(/\[/g, "(").replace(/\]/g, ")");
        nodeLookup[nodeId] = cleanText.length > 50 ? cleanText.slice(0, 50) + "..." : cleanText;
      }

      // Display nodes
      if (nodes.length > 0) {
        results.push("### 📝 Memory Nodes");
        results.push("");
        for (let i = 0; i < Math.min(nodes.length, 10); i++) {
          const node = nodes[i];
          const memory = node.memory ?? "";
          const firstLine = memory.split("\n")[0].slice(0, 100);
          const nodeId = node.id ?? "";
          results.push(`${i + 1}. **${firstLine}**`);
          results.push(`   ID: \`${nodeId}\``);
          results.push("");
        }
      }

      // Display relationships with Mermaid
      if (edges.length > 0) {
        results.push("### 📊 Relationship Diagram (Mermaid)");
        results.push("");
        results.push("```mermaid");
        results.push("graph TD");
        results.push("    classDef cause fill:#f96,stroke:#333,stroke-width:2px;");
        results.push("    classDef relate fill:#bbf,stroke:#333,stroke-width:1px;");
        results.push("    classDef conflict fill:#f66,stroke:#333,stroke-width:2px,stroke-dasharray: 5 5;");

        const addedEdges = new Set<string>();
        for (const edge of edges) {
          const sourceId = edge.source ?? "";
          const targetId = edge.target ?? "";
          const relType = edge.type ?? "UNKNOWN";

          if (relType === "PARENT") continue;

          const edgeKey = `${sourceId}-${targetId}-${relType}`;
          if (addedEdges.has(edgeKey)) continue;
          addedEdges.add(edgeKey);

          const sourceText = nodeLookup[sourceId] ?? sourceId.slice(0, 8);
          const targetText = nodeLookup[targetId] ?? targetId.slice(0, 8);
          const sId = `node_${sourceId.slice(0, 8)}`;
          const tId = `node_${targetId.slice(0, 8)}`;

          if (relType === "CAUSE") {
            results.push(`    ${sId}["${sourceText}"] -- CAUSE --> ${tId}["${targetText}"]:::cause`);
          } else if (relType === "RELATE") {
            results.push(`    ${sId}["${sourceText}"] -. RELATE .- ${tId}["${targetText}"]:::relate`);
          } else if (relType === "CONFLICT") {
            results.push(`    ${sId}["${sourceText}"] == CONFLICT == ${tId}["${targetText}"]:::conflict`);
          } else if (relType === "CONDITION") {
            results.push(`    ${sId}["${sourceText}"] -- CONDITION --> ${tId}["${targetText}"]`);
          } else {
            results.push(`    ${sId}["${sourceText}"] -- ${relType} --> ${tId}["${targetText}"]`);
          }
        }

        results.push("```");
        results.push("");

        // Textual fallback
        results.push("### 🔗 Textual Relationships");
        results.push("");
        results.push("```");
        for (const edge of edges) {
          if (edge.type === "PARENT") continue;
          const sText = (nodeLookup[edge.source ?? ""] ?? "???").slice(0, 40);
          const tText = (nodeLookup[edge.target ?? ""] ?? "???").slice(0, 40);
          results.push(`[${sText}] --${edge.type}--> [${tText}]`);
        }
        results.push("```");
        results.push("");
      }
    }

    results.push("---");
  }

  if (results.length === 0) {
    return "No memories or relationships found.";
  }

  return results.join("\n");
}
